### PR TITLE
[Query Throttler]Gracefully handle config file not found scenario

### PR DESCRIPTION
**Pull request by siddharth16396**
### Description 
- The query throttler config loader previously treated missing config files as errors
- When throttling is not enabled, the config file typically doesn't exist
- This caused unnecessary error logging and potential confusion during normal operation
- The error propagation made it unclear whether throttling was intentionally disabled or misconfigured

### Solution:
- Modified FileBasedConfigLoader.Load() to distinguish between file-not-found and other read errors
- When the config file doesn't exist (os.ErrNotExist), return a default disabled config (Enabled: false)
- This treats a missing file as an intentional "throttling disabled" state rather than an error
- Other file read errors (permissions, I/O errors, etc.) are still propagated as errors
- Updated the test to verify the new behavior returns a disabled config instead of an error

### Test Plan:
- Ran existing unit tests: go test -v ./go/vt/vttablet/tabletserver/querythrottler/...
- Verified the test case "file not found - returns default disabled config" passes
- Confirmed other error scenarios still propagate errors correctly

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by mattlord** [comment]
@siddharth16396 it looks like this introduced a race condition:
```
==================
WARNING: DATA RACE
Read at 0x00c00034b308 by goroutine 12927:
  vitess.io/vitess/go/vt/vttablet/tabletserver/querythrottler.(*QueryThrottler).Throttle()
      /home/runner/work/vitess/vitess/go/vt/vttablet/tabletserver/querythrottler/query_throttler.go:94 +0x70
  vitess.io/vitess/go/vt/vttablet/tabletserver.(*QueryExecutor).Execute()
      /home/runner/work/vitess/vitess/go/vt/vttablet/tabletserver/query_executor.go:161 +0x4cc
  vitess.io/vitess/go/vt/vttablet/tabletserver.(*TabletServer).execute.func1()
      /home/runner/work/vitess/vitess/go/vt/vttablet/tabletserver/tabletserver.go:972 +0x624
  vitess.io/vitess/go/vt/vttablet/tabletserver.(*TabletServer).execRequest()
      /home/runner/work/vitess/vitess/go/vt/vttablet/tabletserver/tabletserver.go:1606 +0xaa1
  vitess.io/vitess/go/vt/vttablet/tabletserver.(*TabletServer).execute()
      /home/runner/work/vitess/vitess/go/vt/vttablet/tabletserver/tabletserver.go:926 +0x309
  vitess.io/vitess/go/vt/vttablet/tabletserver.(*TabletServer).executeWithSettings()
      /home/runner/work/vitess/vitess/go/vt/vttablet/tabletserver/tabletserver.go:1526 +0x1e8
  vitess.io/vitess/go/vt/vttablet/tabletserver.(*TabletServer).ReserveExecute()
      /home/runner/work/vitess/vitess/go/vt/vttablet/tabletserver/tabletserver.go:1435 +0xea
  vitess.io/vitess/go/vt/vttablet/endtoend/framework.(*QueryClient).ReserveExecute()
      /home/runner/work/vitess/vitess/go/vt/vttablet/endtoend/framework/client.go:319 +0x365
  vitess.io/vitess/go/vt/vttablet/endtoend.TestShutdownGracePeriodWithReserveExecute.func2()
      /home/runner/work/vitess/vitess/go/vt/vttablet/endtoend/transaction_test.go:425 +0x56

Previous write at 0x00c00034b308 by goroutine 107:
  vitess.io/vitess/go/vt/vttablet/tabletserver/querythrottler.(*QueryThrottler).startConfigRefreshLoop.func1()
      /home/runner/work/vitess/vitess/go/vt/vttablet/tabletserver/querythrottler/query_throttler.go:163 +0x59a
```

File: go/vt/vttablet/tabletserver/querythrottler/query_throttler.go [hunk 1]
```diff
@@ -84,27 +84,28 @@ func (qt *QueryThrottler) Shutdown() {
...
-// for the significant performance improvement of avoiding mutex contention.
+// This method uses RWMutex read locks to safely access qt.cfg and qt.strategy
+// while allowing concurrent reads from multiple goroutines.
 func (qt *QueryThrottler) Throttle(ctx context.Context, tabletType topodatapb.TabletType, parsedQuery *sqlparser.ParsedQuery, transactionID int64, options *querypb.ExecuteOptions) error {
-	// Lock-free read: for maximum performance in the hot path as cfg and strategy are updated rarely (default once per minute).
-	// They are word-sized and safe for atomic reads; stale data for one query is acceptable and avoids mutex contention in the hot path.
-	if !qt.cfg.Enabled {
+	qt.mu.RLock()
+	cfg := qt.cfg
+	strategy := qt.strategy
```
Comment (mattlord) [inline go/vt/vttablet/tabletserver/querythrottler/query_throttler.go:92]: If this is enough to avoid the race, then I wonder if we can make the strategy an atomic pointer instead?

Comment (siddharth16396) [reply go/vt/vttablet/tabletserver/querythrottler/query_throttler.go:92]: i actually had that as the first change  🤣 
```
siddharth.singh in ~/microrepos/oss/postpone-complete on handle-logs λ git stash list  | cat
stash@{0}: On handle-logs: ATOMIC_POINTERS
stash@{1}: On helpers-attributes: ADDING_METRICS
```

But then i thought it might become an unrelated refactor (which we avoid in uber)
I can raise a seperate PR for that, or i can do the change in this one itself, whatever you guys feel is the right thing

Comment (mattlord) [reply go/vt/vttablet/tabletserver/querythrottler/query_throttler.go:92]: I'm not sure how using a `RWLock` to protect shared memory in order to address the race is more or less of a refactor than using an `atomic.Pointer`? I do now see, however, that there are two variables that we need to protect... I initially thought it was one. So using a `RWLock` may actually be best. 

I would expect to see a meaningful performance impact from this change though and I think it's worth benchmarking it. We may want to consider other alternatives since *I think* this is primarily about eliminating some slightly noisy logging. Or am I missing something?

Thanks, @siddharth16396 !

Comment (siddharth16396) [reply go/vt/vttablet/tabletserver/querythrottler/query_throttler.go:92]: Yes, we need to access two variables, so using an RWLock is the most appropriate approach here.

**Regarding benchmarking:**
This code (with locks in the read-config path) was actually part of the first draft of the Throttler implementation in Uber’s internal fork. The results from that initial version are documented in this RFC: [vitessio/vitess#18412](https://github.com/vitessio/vitess/issues/18412)
.

After that draft, we made a few improvements:

- Lock-free reads in the hot path – this optimized performance significantly but introduced potential data races in certain tests and edge cases.

There were also some internal discussions and follow-up items:

1. Move to using topo server as the config store instead of a file. (This is currently assigned to @stutibiyani, who will raise open-source PRs for it.)

2. Use a watch-based config refresh on the topo server, instead of relying on a background goroutine. (This will be implemented internally first and then open-sourced, along with updated benchmarks.)

Since config update frequency is expected to be extremely low, the overhead of these locks will be negligible once (2) is implemented.

We can either wait for (1) and (2) to be raised as PRs and close this one, or proceed with merging this PR now — both are fine.


Merging now shouldn’t cause any issues because:
- The file-based config reader will currently return Enabled = false, causing the code to exit early due to the continue clause here:

```
newCfg, err := qt.cfgLoader.Load(qt.ctx)
if err != nil {
    log.Errorf("Error loading config: %v", err)
    continue
}
```

Hence, no lock would be taken to write a new config to qt.Cfg.

- The default config refresh frequency is 60 seconds, so even if a config file exists, lock contention would occur only once every 60 seconds.

Comment (mattlord) [reply go/vt/vttablet/tabletserver/querythrottler/query_throttler.go:92]: Thanks, @siddharth16396! With all of that in mind, I'm inclined to say that:
1. We use the log message throttler in v23 (`logutil.NewThrottledLogger()`) if we can, to make it less noisy (it's primarily noisy in aggregate across a cluster of N shards)
2. Hopefully in v24 we have the new and improved method that you outlined above using the topo server with a watch

What do you think?

Comment (mattlord) [reply go/vt/vttablet/tabletserver/querythrottler/query_throttler.go:92]: To be clear, the only reason that I'm being conservative here is that the RWLock overhead (while minimal w/o any active writers) is incurred by every Vitess user on every query, even though most will not have the query throttler enabled. Am I missing or misunderstanding something there? I could be convinced otherwise 🙂

Comment (siddharth16396) [reply go/vt/vttablet/tabletserver/querythrottler/query_throttler.go:92]: Your understanding is totally correct and I agree with you here on both the things:
1) The v23/24 approach is actually right and something even i wish to do.
2) Yes i agree RWLock overhead although minimal will be there and shouldn't ideally be there.

I'll look into using the log message throttler in this PR itself.

One thing which i thought of as a quick stop-gap is:
1) For now i can make this a DEBUG log (assuming DEBUG logs aren't published in planetscale's logging system by default), since right now there's no one using throttler it shouldn't hamper someone debugging and incase they want to debug they can just turn on the debug logs?

What are your thoughts on that?

Comment (mattlord) [reply go/vt/vttablet/tabletserver/querythrottler/query_throttler.go:92]: @siddharth16396 I think that making it a `DEBUG` level log message combined with using the log throttler would be a great combo! ❤️

**Comment by timvaillancourt** [comment]
> When the config file doesn't exist (os.ErrNotExist), return a default disabled config (Enabled: false)

@siddharth16396 I have some concerns about that non-deterministic nature this behaviour introduces. Basically, I feel this approach creates a scenario where the user thinks the throttler is enabled, when in reality it is silently disabled.

A theme in Vitess configuration is if the user asked for X at startup, and for some reason that state cannot be reached, we fail to start the component. Or TL;DR we do exactly what the user asked, and fail if it cannot be achieved.

So I suppose my preference is Vitess components fail to start if we defined a config file that cannot be read/found, vs. silently do something else. That would be more consistent with how configuration works in other components

**Comment by siddharth16396** [comment]
Closed in favor of :: https://github.com/vitessio/vitess/pull/18905
